src/core/border_router/nat64_translator.cpp

```diff
@@ -0,0 +1,454 @@
+        ExitNow(res = Result::kDrop);
+    }
+
+    aMessage.RemoveHeader(sizeof(Ip4::Header));
+
+    ip6Header.Clear();
+    ip6Header.InitVersionTrafficClassFlow();
+    ip6Header.GetSource().SynthesizeFromIp4Address(mNat64Prefix, ip4Header.GetSource());
+    ip6Header.SetDestination(mapping->mIp6);
+    ip6Header.SetFlow(0);
+    ip6Header.SetHopLimit(ip4Header.GetTtl());
+
+    // Note: TCP and UDP are the same for both IPv4 and IPv6 except for the checksum calculation, we will update the
+    // checksum in the payload later. However, we need to translate ICMPv6 packets to ICMP in IPv4 packets.
+    switch (ip4Header.GetProtocol())
+    {
+    case Ip4::kProtoUdp:
+        ip6Header.SetNextHeader(Ip6::kProtoUdp);
+        res = Result::kForward;
+        break;
+    case Ip4::kProtoTcp:
+        ip6Header.SetNextHeader(Ip6::kProtoTcp);
+        res = Result::kForward;
+        break;
+    case Ip4::kProtoIcmp:
+        ip6Header.SetNextHeader(Ip6::kProtoIcmp6);
+        SuccessOrExit(TranslateIcmp4(*mapping, aMessage));
+        res = Result::kForward;
+        break;
+    default:
+        ExitNow(res = Result::kDrop);
+    }
+
+    switch (res)
+    {
+    case Result::kDrop:
+        break;
+    case Result::kReplyIcmp:
+        // TODO: Implement the logic for replying ICMP packets.
+        break;
+    case Result::kForward:
+        ip6Header.SetPayloadLength(aMessage.GetLength() - aMessage.GetOffset());
+        Checksum::UpdateMessageChecksum(aMessage, ip6Header.GetSource(), ip6Header.GetDestination(),
+                                        ip6Header.GetNextHeader());
+        err = aMessage.PrependBytes(&ip6Header, sizeof(ip6Header));
+        if (err != kErrorNone)
+        {
+            // This might happen when the platform failed to reserve enough space before the original IPv4 packet.
+            LogWarn("failed to prepend IPv6 head to translated message");
+            res = Result::kDrop;
+        }
+        break;
+    }
+
+exit:
+    return res;
+}
+
+void Nat64Translator::ReleaseMapping(AddressMapping &aMapping)
+{
+    LogInfo("mapping removed: %s -> %s", aMapping.mIp6.ToString().AsCString(), aMapping.mIp4.ToString().AsCString());
+    mIp4AddressPool[mAvailableAddressCount] = aMapping.mIp4;
+    mAvailableAddressCount++;
+    mAddressMappingPool.Free(aMapping);
+}
+
+Nat64Translator::AddressMapping *Nat64Translator::CreateMapping(const Ip6::Address &aAddr)
+{
+    AddressMapping *mapping = mAddressMappingPool.Allocate();
+
+    if (mapping == nullptr)
+    {
+        uint64_t                   now = Get<Uptime>().GetUptime();
+        LinkedList<AddressMapping> idleMappings;
+
+        mActiveAddressMappings.RemoveAllMatching(now, idleMappings);
+
+        for (AddressMapping *idleMapping = idleMappings.Pop(); idleMapping != nullptr; idleMapping = idleMappings.Pop())
+        {
+            ReleaseMapping(*idleMapping);
+        }
+
+        mapping = mAddressMappingPool.Allocate();
+    }
```

Comment:
Can move the check `VerifyOrExit(mapping != nullptr)` here before the `if (mAvailableAddressCount == 0)` instead of after it.

This way I think we can simplify the `if (mAvailAddressCount == 0)` block. No need to again check for `mapping` against null in the `if` block.

Reply:
Infact, this function can be rewritten into

- Check if we can allocate this mapping an Ip4 address
  - Try to clear the mappings if we don't have available item in the address pool
- Allocate a mapping, and assign the address

This should be correct and safe since the number of available IP4 addresses will always be smaller than the size of mapping pool (since the users may allocate a small CIDR block)